### PR TITLE
#9754: moved split to ttnn

### PR DESCRIPTION
**Pull request by ntarafdar**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/9754)

### Problem description
Moving Split from TT_Lib to TTNN. This op wasn't so widely used so can do this in one PR. 

### What's changed
Replaced TT_Lib.split_last_dim_in_two_chunks and tt_lib.split_any_dim_in_two_chunks with ttnn.split . 
This is pretty analogous to PyTorch split, it just needs to be split in two chunks, support for more chunks will need to be added as well as row-major and sharded support. 

### Checklist
- [x] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/split/device/split_op.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    const auto &input_tensor = input_tensors.at(0);
+    tt::tt_metal::Buffer *in0_buffer = input_tensor.buffer();
+
+    tt::DataFormat cb_data_format = tt::DataFormat::Bfp8_b;
+    if (input_tensor.get_dtype() == tt::tt_metal::DataType::BFLOAT16) {
+        cb_data_format = tt::DataFormat::Float16_b;
+    }
+
+    uint32_t single_tile_size = tt::tt_metal::detail::TileSize(cb_data_format);
+    TT_FATAL(this->dim == 3 || this->dim == 2, "split is possible along dim 2 or 3 only");
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/data_movement/split/device/split_op.cpp:27]:
```suggestion
    TT_FATAL(this->dim == 3 || this->dim == 2, "Split is possible along dim 2 or 3 only");
```

File: ttnn/cpp/ttnn/operations/data_movement/split/device/split_op.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    TT_FATAL(in0_buffer->size() % single_tile_size == 0);
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/data_movement/split/device/split_op.cpp:28]: Could you add comment about buffer size should be divisible by single_tile_size.
If you think it is internal check I suggest to use TT_ASSERT then.

File: ttnn/cpp/ttnn/operations/data_movement/split/device/split_op.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    TT_FATAL(input_tensor.buffer() != nullptr, "Operands to TM need to be allocated in buffers on device!");
+    TT_FATAL(input_tensor.memory_config().memory_layout == TensorMemoryLayout::INTERLEAVED, "Split does not currently support sharding");
+    TT_FATAL(this->output_mem_config.memory_layout == TensorMemoryLayout::INTERLEAVED, "Split does not currently support sharding");
+
+
+    int chunk_size = input_tensor.get_legacy_shape()[this->dim] / this->num_splits;
+    TT_FATAL(input_tensor.get_legacy_shape()[0] == 1, "Only batch 1 implemented");
+    TT_FATAL(input_tensor.get_legacy_shape()[this->dim] % this->num_splits == 0, "Incorrect shape on last dim");
+    TT_FATAL(this->dim <= input_tensor.get_legacy_shape().rank() && this->dim >= 0, "Improper dims");
+    TT_FATAL(input_tensor.get_legacy_shape().rank() == 4, "W,Z,Y,X tensor");
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/data_movement/split/device/split_op.cpp:39]: It is hard to get an explanation.

File: ttnn/cpp/ttnn/operations/data_movement/split/split.hpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    static std::vector<ttnn::Tensor> operator()(
+        uint8_t queue_id,
+        const ttnn::Tensor& input_tensor,
+        const int64_t& num_splits,
+        const int64_t& dim,
+        const std::optional<MemoryConfig>& memory_config_arg);
+
+    static std::vector<ttnn::Tensor> operator()(
+        const ttnn::Tensor& input_tensor,
+        const int64_t& num_splits,
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/data_movement/split/split.hpp:24]: DOnt need to pass int64_t as const ref.

**Review by dmakoviichuk-tt — CHANGES_REQUESTED** [review]
One main request. Please make TT_FATAL comments more actionable or at least provide more details. For example if buffer size cannot be divided by tile size you can provide both sizes. Overall we need to make sure user can understand what he should fix in his code.

File: ttnn/cpp/ttnn/operations/data_movement/split/device/split_op.cpp (hunk 2 above)
Comment (tarafdarTT) [reply ttnn/cpp/ttnn/operations/data_movement/split/device/split_op.cpp:28]: I don't need it (already checking layout), I'll get rid of it. I'll clean this up a bit. I copy pasted from the old tt_lib vsn of it

File: ttnn/cpp/ttnn/operations/data_movement/split/split.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+#include "ttnn/decorators.hpp"
+#include "device/split_op.hpp"
+#include "ttnn/deprecated/tt_dnn/op_library/reshape/reshape_op.hpp"
+#include "ttnn/operations/data_movement/transpose/transpose.hpp"
+#include "ttnn/operations/data_movement/split/split.hpp"
+
+
+namespace ttnn::operations::data_movement {
+
+constexpr uint32_t SplitDefaultQueueID = 0;
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/split/split.cpp:16]: uint8_t

File: tests/ttnn/sweep_tests/sweeps/sweeps/split.py [hunk 6]
```diff
@@ -161,9 +161,7 @@ def run(config, memory_config, *, device) -> Tuple[bool, Optional[str]]:
     ttnn_input_tensor = ttnn.from_torch(
         torch_input_tensor, layout=layout, device=device, dtype=dtype, memory_config=memory_config
     )
-    # TODO: uncomments this when ttnn.split is implemented
-    # ttnn_output_tensors = ttnn.split(ttnn_input_tensor, splits, dim=split_dim)
-    ttnn_output_tensors = ttnn.experimental.tensor.split_dim_two_chunks_tiled(ttnn_input_tensor, split_dim)
+    ttnn_output_tensors = ttnn.split(ttnn_input_tensor, splits, dim=split_dim)
```
Comment (sjameelTT) [inline tests/ttnn/sweep_tests/sweeps/sweeps/split.py:164]: great to see that split sweeps now actually sweeps split

File: ttnn/cpp/ttnn/operations/data_movement/split/split_pybind.hpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                const ttnn::Tensor& input_tensor,
+                int64_t & num_splits,
+                int64_t & dim,
+                const std::optional<ttnn::MemoryConfig>& memory_config,
+                uint8_t queue_id) {
+                    return self(queue_id, input_tensor, num_splits, dim, memory_config);
+                },
+                py::arg("input_tensor"),
+                py::arg("num_splits"),
+                py::arg("dim"),
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/split/split_pybind.hpp:56]: Should we match pytorch and default this to 0?

**Review by TT-BrianLiu — APPROVED** [review]
Minor comments

File: ttnn/cpp/ttnn/operations/data_movement/split/split.cpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    for (Tensor &t : transposed_result) {
+        results.emplace_back(ttnn::transpose(t, dim, 3, mem_config));
+    }
+    return results;
+}
+
+}
+
+
+std::vector<ttnn::Tensor> ExecuteSplit::operator()(
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/data_movement/split/split.cpp:67]: SplitOperation instead of ExecuteSplit is what we're going with now

**Review by dmakoviichuk-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/split/device/split_op.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+#include "tt_metal/host_api.hpp"
+#include "tt_metal/common/constants.hpp"
+
+#include "split_program_factory.hpp"
+using namespace tt::constants;
+
+
+namespace ttnn::operations::data_movement {
+
+void Split::validate(const std::vector<Tensor> &input_tensors) const {
```
Comment (sjameelTT) [inline ttnn/cpp/ttnn/operations/data_movement/split/device/split_op.cpp:17]: SplitDeviceOperation

File: tests/tt_eager/python_api_testing/sweep_tests/tt_lib_ops.py [hunk 10]
```diff
@@ -1790,10 +1790,10 @@ def reshape(
 @setup_host_and_device
 def split_last_dim_two_chunks_tiled(x, *args, device, dtype, layout, input_mem_config, output_mem_config, **kwargs):
     t0 = setup_tt_tensor(x, device, layout[0], input_mem_config[0], dtype[0])
-    t1 = ttl.tensor.split_last_dim_two_chunks_tiled(t0, output_mem_config=output_mem_config)
+    t1 = ttnn.split(t0, 2, 3, memory_config=output_mem_config)
 
-    output0 = tt2torch_tensor(t1[0])
-    output1 = tt2torch_tensor(t1[1])
+    output0 = t1[0].cpu().to_torch()
```
Comment (cfjchu) [inline tests/tt_eager/python_api_testing/sweep_tests/tt_lib_ops.py:1795]: I recommend this instead:
```suggestion
    output0 = ttnn.to_torch(t1[0])
```

File: ttnn/cpp/ttnn/operations/data_movement/split/device/split_program_factory.hpp [hunk 11]
```diff
@@ -0,0 +1,21 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttnn/deprecated/tt_dnn/op_library/work_split.hpp"
```
Comment (cfjchu) [inline ttnn/cpp/ttnn/operations/data_movement/split/device/split_program_factory.hpp:5]: we probably don't need to include all these headers here, right?

**Review by arakhmati — APPROVED** [review]

**Review by cfjchu — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

**Review by tt-rkim — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]

**Pull request merged**